### PR TITLE
Enhancements in working with objects

### PR DESCRIPTION
**Pull request by skletsun**
## Overview

- `JNIEnv` methods are updated to accept anything that is convertible into `JObject` (used to accept `JObject` only). This should not break existing client code.
- Implemented `JNIEnv#is_same_object()`. See https://docs.oracle.com/en/java/javase/11/docs/specs/jni/functions.html#issameobject

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has documentation
- [x] This change is not breaking **or** mentioned in the Changelog
- [x] The [continuous integration build](https://www.travis-ci.org/jni-rs/jni-rs) passes

### CODE REVIEW

File: tests/jni_api.rs [hunk 1]
```diff
@@ -600,13 +590,33 @@ pub fn throw_defaults() {
     test_throwable_descriptor_with_default_type(&env, JNIString::from(TEST_EXCEPTION_MESSAGE));
 }
 
+#[test]
+pub fn test_conversion() {
```
Comment (dmitry-timofeev) [inline tests/jni_api.rs]: Shan't this be tested more explicitly (it is unclear what the string length has to do with the conversion):

```
JObject src = <any object, let it be same string>
T wrapper = create_wrapper(src);
JObject actual = wrapper.into();
assert_that(is_same_instance(&env, src, actual));
```

File: src/wrapper/jnienv.rs [hunk 2]
```diff
@@ -917,15 +932,23 @@ impl<'a> JNIEnv<'a> {
     /// Cast a JObject to a `JList`. This won't throw exceptions or return errors
     /// in the event that the object isn't actually a list, but the methods on
     /// the resulting map object will.
-    pub fn get_list(&self, obj: JObject<'a>) -> Result<JList<'a, '_>> {
+    pub fn get_list<O>(&self, obj: O) -> Result<JList<'a, '_>>
```
[the remark this answers is not in the document]
Comment (skletsun) [reply src/wrapper/jnienv.rs]: Perhaps it's better to roll it back for `.get_list()` and `.det_map()`. If a user wants to wrap it with "smart" wrappers better to do everything explicitly.

File: tests/jni_api.rs [hunk 3]
```diff
@@ -592,22 +592,35 @@ pub fn throw_defaults() {
...
 pub fn test_conversion() {
-    fn test_str_len<'a: 'b, 'b>(env: &JNIEnv<'a>, obj: impl Into<JObject<'b>>) {
-        let len = env
-            .call_method(obj, "length", "()I", &[])
+    fn assert_is_same_object<'a: 'b, 'b>(
+        env: &JNIEnv<'a>,
+        src: JObject<'b>,
+        actual: impl Into<JObject<'b>>,
+    ) {
+        let equals = env
```
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply tests/jni_api.rs]: Also, shall the test be more explicit with into, as was suggested, for it is the operation under test (i.e., convert in the test)?

[the remark this answers is not in the document]
Comment (skletsun) [reply tests/jni_api.rs]: > It seems we don't support https://docs.oracle.com/en/java/javase/11/docs/specs/jni/functions.html#issameobject 🙀 (I suggested to use that)
> 
> We can add it separately, I think.

Implemented it within this PR - if it looks OK I'd prefer not to separate it.

File: tests/jni_api.rs [hunk 4]
```diff
@@ -101,6 +101,19 @@ pub fn is_instance_of_null() {
...
+#[test]
+pub fn test_same_objects() {
+    const SOURCE_STR: &str = "THIS IS STRING OBJECT";
+    let env = attach_current_thread();
+    let string = env.new_string(SOURCE_STR).unwrap();
+    let ref_from_string = unwrap(&env, env.new_local_ref::<JObject>(string.into()));
+    assert!(unwrap(&env, env.is_same_object(string, ref_from_string)));
+    unwrap(&env, env.delete_local_ref(ref_from_string));
+
+    let same_src_str = env.new_string(SOURCE_STR).unwrap();
```
Comment (dmitry-timofeev) [inline tests/jni_api.rs]: These are different test-cases, I think it is appropriate to put them into separate test methods, e.g.:
- test_same_object_diff_references, 
- test_same_object_same_reference
- test_not_same_object
- test_not_same_object_null

File: src/wrapper/jnienv.rs [hunk 5]
```diff
@@ -180,6 +180,21 @@ impl<'a> JNIEnv<'a> {
         ) == sys::JNI_TRUE)
     }
 
+    /// Returns true if ref1 and ref2 refer to the same Java object, or are both `NULL`. Otherwise,
+    /// returns false.
+    pub fn is_same_object<'b, 'c, O, T>(&self, ref1: O, ref2: T) -> Result<bool>
```
Comment (dmitry-timofeev) [inline src/wrapper/jnienv.rs]: It could also extend the changelog.

**Comment by dmitry-timofeev** [comment]
👍 , thanks! 

I think it might make sense to release some of the patches.

**Pull request merged**